pkg/controller/vsphere/actuator_test.go

```diff
@@ -0,0 +1,287 @@
+			},
+			event: "test: reconciler failed to Update machine: test: failed validating machine provider spec: test: missing \"machine.openshift.io/cluster-api-cluster\" label",
+		},
+		{
+			name: "Update machine event succeed and only one event is created",
+			operation: func(actuator *Actuator, machine *machinev1.Machine) {
+				actuator.Update(context.Background(), machine)
+				actuator.Update(context.Background(), machine)
+			},
+			event: "Updated Machine test",
+		},
+		{
+			name: "Delete machine event failed on invalid machine scope",
+			operation: func(actuator *Actuator, machine *machinev1.Machine) {
+				actuator.Delete(nil, machine)
+			},
+			event: "test: failed to create scope for machine: test: machine scope require a context",
+		},
+		{
+			name: "Delete machine event failed, reconciler's delete failed",
+			operation: func(actuator *Actuator, machine *machinev1.Machine) {
+				actuator.Delete(context.Background(), machine)
+			},
+			event: "test: reconciler failed to Delete machine: destroying vm in progress, reconciling",
+		},
+		{
+			name: "Delete machine event succeed",
+			operation: func(actuator *Actuator, machine *machinev1.Machine) {
+				actuator.Delete(context.Background(), machine)
+			},
+			event: "Deleted machine test",
+		},
+	}
+
+	for _, tc := range cases {
+		t.Run(tc.name, func(t *testing.T) {
+			ctx := context.Background()
+			timeout := 10 * time.Second
+			gs := NewWithT(t)
+
+			providerSpec, err := vsphereapi.RawExtensionFromProviderSpec(&vsphereapi.VSphereMachineProviderSpec{
+				Template: vm.Name,
+				Workspace: &vsphereapi.Workspace{
+					Server: host,
+				},
+				CredentialsSecret: &corev1.LocalObjectReference{
+					Name: "test",
+				},
+			})
+			if err != nil {
+				t.Fatal(err)
+			}
+
+			machine := &machinev1.Machine{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "test",
+					Namespace: "test",
+					Labels: map[string]string{
+						machinev1.MachineClusterIDLabel: "CLUSTERID",
+					},
+				},
+				Spec: machinev1.MachineSpec{
+					ProviderSpec: machinev1.ProviderSpec{
+						Value: providerSpec,
+					},
+				},
+			}
+
+			// Create the machine
+			gs.Expect(k8sClient.Create(ctx, machine)).To(Succeed())
+			defer func() {
+				gs.Expect(k8sClient.Delete(ctx, machine)).To(Succeed())
+			}()
+
+			// Ensure the machine has synced to the cache
+			getMachine := func() error {
+				machineKey := types.NamespacedName{Namespace: machine.Namespace, Name: machine.Name}
+				return k8sClient.Get(ctx, machineKey, machine)
+			}
+			gs.Eventually(getMachine, timeout).Should(Succeed())
+
+			params := ActuatorParams{
+				Client:        k8sClient,
+				EventRecorder: eventRecorder,
+				APIReader:     k8sClient,
+			}
+
+			actuator := NewActuator(params)
+			tc.operation(actuator, machine)
+
+			eventList := &v1.EventList{}
+			waitForEvent := func() error {
+				gs.Expect(k8sClient.List(ctx, eventList, client.InNamespace(machine.Namespace))).To(Succeed())
```

Comment:
Better to not use gomega inside functions that are used in eventually blocks, otherwise if the list fails, the whole test will fail rather than retrying
```suggestion
				err := k8sClient.List(ctx, eventList, client.InNamespace(machine.Namespace))
				if err != nil {
				  return err
				}
```